src/Raven.Server/Documents/CountersStorage.cs

```diff
@@ -163,162 +173,598 @@ public long GetNumberOfCountersToProcess(DocumentsOperationContext context, stri
                         }
-                    }
 
-                    RemoveTombstoneIfExists(context, documentId, name);
+                        if (data == null)
+                        {
+                            data = context.ReadObject(sourceData, null, BlittableJsonDocumentBuilder.UsageMode.ToDisk);
+                            _countersCount += sourceCounters?.Count ?? 0;
+                        }
 
-                    var etag = _documentsStorage.GenerateNextEtag();
+                        var etag = _documentsStorage.GenerateNextEtag();
 
-                    if (changeVector == null)
-                    {
-                        changeVector = ChangeVectorUtils
-                            .TryUpdateChangeVector(_documentDatabase.ServerStore.NodeTag, _documentsStorage.Environment.Base64Id, etag, string.Empty)
-                            .ChangeVector;
+                        if (changeVector == null)
+                        {
+                            changeVector = ChangeVectorUtils
+                                .TryUpdateChangeVector(_documentDatabase.ServerStore.NodeTag, _documentsStorage.Environment.Base64Id, etag, string.Empty)
+                                .ChangeVector;
+                        }
+
+                        using (Slice.From(context.Allocator, changeVector, out var cv))
+                        using (DocumentIdWorker.GetStringPreserveCase(context, collectionName.Name, out Slice collectionSlice))
+                        {
+                            tvb.Add(counterKey);
+                            tvb.Add(Bits.SwapBytes(etag));
+                            tvb.Add(cv);
+                            tvb.Add(data.BasePointer, data.Size);
+                            tvb.Add(collectionSlice);
+                            tvb.Add(context.TransactionMarkerOffset);
+
+                            table.Set(tvb);
+                        }
                     }
+                }
+            }
+            finally 
+            {
+                foreach (var s in _counterModificationMemoryScopes)
+                {
+                    s.Dispose();
+                }
+                _counterModificationMemoryScopes.Clear();              
+            }
+        }
+
+        private static List<LazyStringValue> DbIdsToList(BlittableJsonReaderArray dbIds)
+        {
+            var localDbIdsList = new List<LazyStringValue>(dbIds.Length);
+            for (int i = 0; i < dbIds.Length; i++)
+            {
+                localDbIdsList.Add((LazyStringValue)dbIds[i]);
+            }
 
-                    using (Slice.From(context.Allocator, changeVector, out var cv))
-                    using (DocumentIdWorker.GetStringPreserveCase(context, collectionName.Name, out Slice collectionSlice))
+            return localDbIdsList;
+        }
+
+        private long InternalPutCounter(DocumentsOperationContext context, BlittableJsonReaderObject counters, string counterName,
```

Comment:
`localDbIds` - extract to a struct that contains both the array and the list in one shot. And can apply operations to them

Reply:
Should return `bool` here

Reply:
done